File: api/internal/domain/models/game_charge.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type GameCharge struct {
	ID           string              `json:"id"`
	GameID       string              `json:"gameId"`
	UserID       *string             `json:"userId"`       // Nullable for unclaimed charges
	ExternalName string              `json:"externalName"` // Name from spreadsheet if UserID is null
	AmountCents  int                 `json:"amountCents"`
	CreatedAt    time.Time           `json:"createdAt"`
	Allocations  []PaymentAllocation `json:"allocations"`
	Game         *Game               `json:"game,omitempty"`
}

func (c GameCharge) IsPaid() bool {
	paid := 0
	for _, a := range c.Allocations {
		paid += a.AmountInCents
	}
	return paid >= c.AmountCents
}

func CreateGameCharge(gameID string, userID *string, externalName string, amount int) *GameCharge {
	id, _ := uuid.NewV7()
	return &GameCharge{
		ID:           id.String(),
		GameID:       gameID,
		UserID:       userID,
		ExternalName: externalName,
		AmountCents:  amount,
		CreatedAt:    time.Now(),
		Allocations:  make([]PaymentAllocation, 0),
	}
}

File: bouncy-ui/lib/models/game_charge.dart
import 'package:ui_test/models/payment_allocation.dart';

class GameCharge {
  final String id;
  final String gameId;
  final String userId;
  final int amountCents;
  final DateTime createdAt;
  final List<PaymentAllocation> allocations;

  GameCharge({
    required this.id,
    required this.gameId,
    required this.userId,
    required this.amountCents,
    required this.createdAt,
    required this.allocations,
  });

  factory GameCharge.fromMap(Map<String, dynamic> map) {
    return GameCharge(
      id: map['ID'] ?? map['id'] ?? '',
      gameId: map['GameID'] ?? map['gameId'] ?? map['game_id'] ?? '',
      userId: map['UserID'] ?? map['userId'] ?? map['user_id'] ?? '',
      amountCents: map['AmountCents'] ?? map['amountCents'] ?? 0,
      createdAt:
          DateTime.tryParse(map['CreatedAt'] ?? map['createdAt'] ?? '') ??
              DateTime.now(),
      allocations: List<PaymentAllocation>.from(
        ((map['Allocations'] ?? map['allocations']) ?? [])
            .map((x) => PaymentAllocation.fromMap(x)),
      ),
    );
  }
}
